core/src/nucleus/ribosome/api/query.rs

```diff
@@ -59,33 +69,88 @@ pub fn invoke_query(runtime: &mut Runtime, args: &RuntimeArgs) -> ZomeApiResult
+                },
             )
         }
         QueryArgsNames::QueryName(name) => {
             let refs: Vec<&str> = vec![&name]; // String -> Vec<&str>
             agent.chain().query(
                 &Some(top),
                 refs.as_slice(), // Vec<&str> -> &[&str]
-                query.start,
-                query.limit,
+                ChainStoreQueryOptions {
+                    start: query.options.start,
+                    limit: query.options.limit,
+                    headers: query.options.headers,
+                },
             )
         }
     };
-    let result = match addresses {
+    let result = match maybe_result {
         // TODO #793: the Err(_code) is the RibosomeErrorCode, but we can't import that type here.
         // Perhaps return chain().query should return Some(result)/None instead, and the fixed
         // UnknownEntryType code here, rather than trying to return a specific error code.
-        Ok(result) => Ok(result),
+        Ok(result) => Ok(match (query.options.entries, result) {
+            (false, ChainStoreQueryResult::Addresses(addresses)) => {
+                QueryResult::Addresses(addresses)
+            }
+            (false, ChainStoreQueryResult::Headers(headers)) => QueryResult::Headers(headers),
+            (true, ChainStoreQueryResult::Addresses(addresses)) => {
+                let maybe_entries: Result<Vec<(Address,Entry)>,HolochainError> = addresses
+                    .iter()
+                    .map(|address| // -> Result<Entry, HolochainError>
+                         Ok((address.to_owned(), get_entry_from_context(&runtime.context, address)?)))
+                    .collect();
+
+                match maybe_entries {
+                    Ok(entries) => QueryResult::Entries(entries),
+                    Err(_e) => return ribosome_error_code!(UnknownEntryType), // TODO: return actual error?
+                }
+            }
+            (true, ChainStoreQueryResult::Headers(headers)) => {
+                let maybe_headers_with_entries: Result<Vec<(ChainHeader,Entry)>,HolochainError> = headers
+                    .iter()
+                    .map(|header| // -> Result<Entry, HolochainError>
+                         Ok((header.to_owned(), get_entry_from_context(&runtime.context,header.entry_address())?)))
+                    .collect();
+                match maybe_headers_with_entries {
+                    Ok(headers_with_entries) => {
+                        QueryResult::HeadersWithEntries(headers_with_entries)
+                    }
+                    Err(_e) => return ribosome_error_code!(UnknownEntryType), // TODO: return actual error?
+                }
+            }
+        }),
         Err(_code) => return ribosome_error_code!(UnknownEntryType),
     };
 
     runtime.store_result(result)
 }
+
+/// Get an Entry via the provided context, returning Entry or HolochainError on failure
+fn get_entry_from_context(
```

Comment:
Isn't query supposed to operate on the chain only? Why `from_context` which actually is do a get `from_dht`? This only works because we currently reuse the same CAS for both, chain and dht, but we do have to separate CASes in the context to keep them separate. This function should be `get_entry_from_chain` which only operates on the source chain.

Reply:
Done.  I've augmented core/src/nucleus/actions/get_entry.rs to provide crate-local APIs to access Entry values from either the either the context's dht or the agent.

Reply:
👍